src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/OAuth2/OAuthClient.cs

```diff
@@ -0,0 +1,255 @@
+                return await HandleDeviceAuthChallengeAsync<T>(ex.Response).ConfigureAwait(false);
+            }
+
+            return typedResponse;
+        }
+
+        private bool IsDeviceAuthChallenge(IHttpWebResponse response, bool respondToDeviceAuthChallenge)
+        {
+            return DeviceAuthHelper.CanHandleDeviceAuthChallenge
+                   && response != null
+                   && respondToDeviceAuthChallenge
+                   && response?.Headers != null
+                   && response.StatusCode == HttpStatusCode.Unauthorized
+                   && response.Headers.Contains(WwwAuthenticateHeader)
+                   && response.Headers.GetValues(WwwAuthenticateHeader).FirstOrDefault()
+                       .StartsWith(PKeyAuthName, StringComparison.OrdinalIgnoreCase);
+        }
+
+        private IDictionary<string, string> ParseChallengeData(IHttpWebResponse response)
+        {
+            IDictionary<string, string> data = new Dictionary<string, string>();
+            string wwwAuthenticate = response.Headers.GetValues(WwwAuthenticateHeader).FirstOrDefault();
+            wwwAuthenticate = wwwAuthenticate.Substring(PKeyAuthName.Length + 1);
+            List<string> headerPairs = EncodingHelper.SplitWithQuotes(wwwAuthenticate, ',');
+            foreach (string pair in headerPairs)
+            {
+                List<string> keyValue = EncodingHelper.SplitWithQuotes(pair, '=');
+                data.Add(keyValue[0].Trim(), keyValue[1].Trim().Replace("\"", ""));
+            }
+
+            return data;
+        }
+
+        private async Task<T> HandleDeviceAuthChallengeAsync<T>(IHttpWebResponse response)
+        {
+            IDictionary<string, string> responseDictionary = ParseChallengeData(response);
+
+            if (!responseDictionary.ContainsKey("SubmitUrl"))
+            {
+                responseDictionary["SubmitUrl"] = RequestUri;
+            }
+
+            string responseHeader = await DeviceAuthHelper.CreateDeviceAuthChallengeResponseAsync(responseDictionary)
+                .ConfigureAwait(false);
+            IRequestParameters rp = BodyParameters;
+            CheckForExtraQueryParameter(responseDictionary["SubmitUrl"]);
+            BodyParameters = rp;
+            _headers["Authorization"] = responseHeader;
+            return await GetResponseAsync<T>(false).ConfigureAwait(false);
+        }
+
+        private static string CheckForExtraQueryParameter(string url)
+        {
+            string extraQueryParameter = PlatformProxyFactory.GetPlatformProxy().GetEnvironmentVariable(ExtraQueryParamEnvVariable);
+            string delimiter = (url.IndexOf('?') > 0) ? "&" : "?";
+            if (!string.IsNullOrWhiteSpace(extraQueryParameter))
+            {
+                url += string.Concat(delimiter, extraQueryParameter);
+            }
+
+            return url;
+        }
+
+        internal async Task<IHttpWebResponse> ExecuteRequestAsync<T>()
```

Comment:
So this class exposes 2 public methods that do almost the same thing - one that executes a request and serializes to an object T, and one that just executes the request. I am wondering if the second one is needed.